src/kcmd/grub.go

```diff
@@ -5,84 +5,85 @@ import (
 	"fmt"
 	"log"
 	"os"
-	"sort"
 	"strings"
 
 	"github.com/canonical/rt-conf/src/model"
+	"github.com/canonical/rt-conf/src/utils"
 )
 
-func sortKcmdlineParams(cmdline string) string {
-	params := strings.Fields(cmdline)
-	sort.Strings(params)
-	return strings.Join(params, " ")
-}
-
 // UpdateGrub reads GRUB_CMDLINE_LINUX_DEFAULT from the default GRUB configuration file,
 // merges it with the kernel command line parameters specified in the provided config,
 // and writes the resulting command line to a drop-in configuration file for GRUB.
 func UpdateGrub(cfg *model.InternalConfig) ([]string, error) {
-
-	params := model.ConstructKeyValuePairs(&cfg.Data.KernelCmdline)
-	if len(params) == 0 {
+	if len(cfg.Data.KernelCmdline.Parameters) == 0 {
 		return nil, fmt.Errorf("no parameters to inject")
 	}
 
-	cmdline, err := parseGrubCMDLineLinuxDefault(cfg.GrubCfg.GrubDefaultFilePath)
+	// Convert new parameters to Params for easier manipulation
+	newParams := cfg.Data.KernelCmdline.ToParams()
+
+	// Parse existing GRUB command line
+	existingCmdlineStr, err := parseGrubCMDLineLinuxDefault(cfg.GrubCfg.GrubDefaultFilePath)
 	if err != nil {
-		return nil, fmt.Errorf("failed to parse %s grub file: %s",
-			cfg.GrubCfg.GrubDefaultFilePath, err.Error())
+		return nil, fmt.Errorf("failed to parse %s grub file: %v", cfg.GrubCfg.GrubDefaultFilePath, err)
 	}
 
-	if err := duplicatedParams(cmdline); err != nil {
-		return nil, fmt.Errorf(
-			"invalid existing parameters in %s for GRUB_CMDLINE_LINUX_DEFAULT: %s",
-			cfg.GrubCfg.GrubDefaultFilePath, err)
+	// Create KernelCmdline from existing string and validate
+	existingCmdline := model.NewKernelCmdline(existingCmdlineStr)
+	if err := existingCmdline.HasDuplicates(); err != nil {
+		return nil, fmt.Errorf("invalid existing parameters in %s: %v", cfg.GrubCfg.GrubDefaultFilePath, err)
 	}
-	currParams := model.CmdlineToParams(cmdline)
 
-	// This replaces if the param already exists and
-	// creates a new one if it doesn't
-	for k, v := range params {
-		currParams[k] = v
-	}
+	// Convert existing cmdline to params
+	existingParams := existingCmdline.ToParams()
+
+	// Merge parameters (new ones override existing ones)
+	existingParams.Merge(newParams)
 
-	cfg.GrubCfg.Cmdline = model.ParamsToCmdline(currParams)
-	cfg.GrubCfg.Cmdline = sortKcmdlineParams(cfg.GrubCfg.Cmdline)
-	cmdline = sortKcmdlineParams(cmdline)
+	// Convert back to sorted command line string
+	cfg.GrubCfg.Cmdline = model.ParamsToCmdline(existingParams)
+
+	// Sort both command lines for consistent comparison
+	existingCmdlineStr = model.ParamsToCmdline(existingCmdline.ToParams())
```

Comment:
I don't see how this adds consistency:
```
Default kernel command line:
-  quiet splash
New kernel command line:
+  irqaffinity=0-1 isolcpus=2-3 kthread_cpus=0-1 nohz=on nohz_full=2-3 quiet rcu_nocbs=0-1 splash
```

I would rather expect:
```
Default kernel command line:
-  quiet splash
New kernel command line:
+  quiet splash irqaffinity=0-1 isolcpus=2-3 kthread_cpus=0-1 nohz=on nohz_full=2-3 rcu_nocbs=0-1
```
With `isolcpus=2-3 kthread_cpus=0-1 nohz=on nohz_full=2-3 rcu_nocbs=0-1` sorted exactly same as the user input.